kernel/kernel-auth-adapter/src/main/java/io/mosip/kernel/auth/defaultadapter/config/SelfTokenRestInterceptor.java

```diff
@@ -0,0 +1,169 @@
+import org.springframework.web.client.HttpServerErrorException;
+import org.springframework.web.client.RestTemplate;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import io.mosip.kernel.auth.defaultadapter.constant.AuthAdapterConstant;
+import io.mosip.kernel.auth.defaultadapter.constant.AuthAdapterErrorCode;
+import io.mosip.kernel.auth.defaultadapter.exception.AuthAdapterException;
+import io.mosip.kernel.auth.defaultadapter.exception.AuthRestException;
+import io.mosip.kernel.auth.defaultadapter.util.CachedTokenObject;
+import io.mosip.kernel.core.authmanager.model.ClientSecret;
+import io.mosip.kernel.core.exception.ExceptionUtils;
+import io.mosip.kernel.core.exception.ServiceError;
+import io.mosip.kernel.core.http.RequestWrapper;
+
+/**
+ * This class intercepts and renew client token.
+ * 
+ * @author Urvil Joshi
+ *
+ */
+public class SelfTokenRestInterceptor implements ClientHttpRequestInterceptor {
+
+	private String clientID;
+
+	private String clientSecret;
+
+	private String appID;
+
+	private String tokenURL;
+	
+	private String validateTokenURL;
+
+	private CachedTokenObject<String> cachedToken;
+
+	private static final Logger LOGGER = LoggerFactory.getLogger(SelfTokenRestInterceptor.class);
+
+	private RestTemplate restTemplate;
+
+	public SelfTokenRestInterceptor(Environment environment, RestTemplate restTemplate,
+			CachedTokenObject<String> cachedToken) {
+		clientID = environment.getProperty("mosip.iam.adapter.clientid", "");
+		clientSecret = environment.getProperty("mosip.iam.adapter.clientsecret", "");
+		appID = environment.getProperty("mosip.iam.adapter.appid", "");
+		tokenURL = environment.getProperty("mosip.authmanager.client-token-endpoint", "");
+		validateTokenURL = environment.getProperty("auth.server.admin.validate.url", "");
+		this.cachedToken = cachedToken;
+		this.restTemplate = restTemplate;
+	}
+
+	ObjectMapper objectMapper = new ObjectMapper();
+
+	@Override
+	public ClientHttpResponse intercept(HttpRequest request, byte[] body, ClientHttpRequestExecution execution)
+			throws IOException {
+		// null check if job is not able to fetch client id secret
+		if (cachedToken.getToken() == null) {
+			LOGGER.error("there is some issue with getting token with clienid and secret");
+		}
+		request.getHeaders().add(AuthAdapterConstant.AUTH_HEADER_COOKIE,
+				AuthAdapterConstant.AUTH_HEADER + cachedToken.getToken());
+
+		ClientHttpResponse clientHttpResponse = execution.execute(request, body);
+		if(clientHttpResponse.getStatusCode() != HttpStatus.UNAUTHORIZED) {
+			return clientHttpResponse;
+		}
+		synchronized (this) {
+			// online validation
+			if(!isTokenValid(cachedToken.getToken())) {
+			String authToken = getClientToken(clientID, clientSecret, appID);
+			cachedToken.setToken(authToken);		
+			}
+		}
+		
+		List<String> cookies = request.getHeaders().get(AuthAdapterConstant.AUTH_HEADER_COOKIE);
+		if (cookies != null && !cookies.isEmpty()) {
```

Comment:
why are we doing this cookie collect and putting it back ?

Reply:
we are removing old cookies.